plugins/jabra-file/fu-jabra-file-device.c

```diff
@@ -0,0 +1,855 @@
+	g_autoptr(FuFirmware) firmware = fu_jabra_file_firmware_new();
+
+	if (!fu_firmware_parse_stream(firmware, stream, 0x0, flags, error))
+		return NULL;
+	if (fu_jabra_file_firmware_get_dfu_pid(FU_JABRA_FILE_FIRMWARE(firmware)) != self->dfu_pid) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_DATA,
+			    "wrong DFU PID, got 0x%x, expected 0x%x",
+			    fu_jabra_file_firmware_get_dfu_pid(FU_JABRA_FILE_FIRMWARE(firmware)),
+			    self->dfu_pid);
+		return NULL;
+	}
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_jabra_file_device_setup(FuDevice *device, GError **error)
+{
+	FuJabraFileDevice *self = FU_JABRA_FILE_DEVICE(device);
+	if (!fu_jabra_file_device_read_name(self, error))
+		return FALSE;
+	if (!fu_jabra_file_device_read_version(self, error))
+		return FALSE;
+	if (!fu_jabra_file_device_read_dfu_pid(self, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_file_device_write_firmware(FuDevice *device,
+				    FuFirmware *firmware,
+				    FuProgress *progress,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	FuJabraFileDevice *self = FU_JABRA_FILE_DEVICE(device);
+	const guint chunk_size = 55;
+	g_autoptr(FuChunkArray) chunks = NULL;
+	g_autoptr(GInputStream) stream = NULL;
+	gchar *firmware_checksum = NULL;
+	GBytes *firmware_bytes = NULL;
+	gboolean match = FALSE;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 1, "first_block");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 89, "next_block");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 10, "update_device");
+
+	/* check if firmware file already exists on device */
+	firmware_bytes = fu_firmware_get_bytes(firmware, error);
+	firmware_checksum = g_compute_checksum_for_bytes(G_CHECKSUM_MD5, firmware_bytes);
+	if (!fu_jabra_file_device_file_checksum(self, firmware_checksum, &match, error)) {
+		return FALSE;
+	}
+
+	if (!match) {
+		/* file not on device, transfer it */
+		stream = fu_firmware_get_stream(firmware, error);
+		if (stream == NULL)
+			return FALSE;
+		chunks = fu_chunk_array_new_from_stream(stream, 0x00, chunk_size, error);
+		if (chunks == NULL)
+			return FALSE;
+
+		if (!fu_jabra_file_device_write_delete_file(self, error))
+			return FALSE;
+		if (!fu_jabra_file_device_write_first_block(self, firmware, error))
+			return FALSE;
+		fu_progress_step_done(progress);
+		if (!fu_jabra_file_device_write_blocks(self,
+						       chunks,
+						       fu_progress_get_child(progress),
+						       error))
+			return FALSE;
+		fu_progress_step_done(progress);
+		match = FALSE;
+		if (!fu_jabra_file_device_file_checksum(self, firmware_checksum, &match, error)) {
+			return FALSE;
+		}
+		if (!match)
+			return FALSE;
+	} else {
+		/* file already on device, adjust progress */
+		fu_progress_step_done(progress);
+		fu_progress_step_done(progress);
+	}
```

Comment:
i don't get this bit -- if you "reinstall" a firmware to test the plugin you're not actually going to test anything. It's a bit odd.

Reply:
This is a protocol specific thing, and not really a test. Basically since the transfer can take a while (it's a rather large firmware file, using a rather slow transfer method), if the device is busy, the file is not removed from the device on a failure, meaning if trying to update again, the file is already on the device, hence there's no need to do the slow transfer again. 
This is how I thought to best reflect this on the progress, but there could be different ways, I would guess

Reply:
> This is how I thought to best reflect this on the progress

I think split up the (!match) case into a different function block and pass it a child progress. Then you can do a different fu_progress_set_id() block for the "already exists" case. If you're completely lost say and I'll commit something that might make sense.

Reply:
I think I understand the idea! I'll give it a shot.
just an fyi, I'm currently on a slightly reduced work schedule, so it may take me a bit longer to go through all the changes in the PR

Reply:
No problem at all -- no panic from my side.